### DISCUSSION TITLE
Support for switching channels at media_player

### DISCUSSION THREAD

**Discussion by mib1185** [home-assistant/architecture#515] [Entity Models]
# Context
At least the following integrations use of `media_next_track` and `media_previous_track` services to switch channels
- androidtv
- aquostv
- braviatv
- mediaroom
- samsungtv
- webostv

All of them (_except androidtv_) implements the STATE_ON (_but not STATE_PLAYING_) which honestly is the more correct one for suggesting the TV is on (_and is not playing any video on demand or other non linear media content_).
Prior to changes in home-assistant/frontend#7389 for entities of these integrations the media_player card was showing the next and previous track buttons, to switch the TV channels.
To be honest, there's a difference between playing a track and switching, and watching the linear TV channels and switching. 

# Proposal
Add two additional services `media_next_channel` and `media_previous_channel` to media_player integration and rework the TV integrations to use these ones, instead of next and prev. track services, to switch the TV channels.

Suggested "core" changes:
- core: https://github.com/mib1185/homeassistant-core/commit/e5a4d26baa6c6327e2abe65da50ed06157a8f8c4
- frontend: https://github.com/mib1185/homeassistant-frontend/commit/73c4ac99af39ffb6ef5a1585772425299607b84d

Suggest integration changes:
- androidtv: _tbd_
- aquostv: https://github.com/mib1185/homeassistant-core/commit/8131ac92a9c889b880177dee38a052d15bc79e77
- braviatv: may need additional work in dependency lib
- mediaroom: https://github.com/mib1185/homeassistant-core/commit/f25be84c63b976bfea5b4b8c39d912cae98222d0 (breaking change)
- samsungtv: https://github.com/mib1185/homeassistant-core/commit/35c19a0e3dccb5dddf8ae29a41f54f252319b0c9 (breaking change)
- webostv: https://github.com/mib1185/homeassistant-core/commit/c1e159f555e67cc9ffed25f018c7b43773bae3af (breaking change)

Preview:
![image](https://user-images.githubusercontent.com/35783820/132987985-e519fc1b-cc28-4e0a-9449-01c35d1f1b96.png)

![image](https://user-images.githubusercontent.com/35783820/132987995-6e1c7d4f-d14a-4a35-a342-7ee5be50662a.png)

# Consequences
At least the following integrations should be updated accordingly:
- aquostv
- braviatv
- mediaroom
- samsungtv
- webostv

**Comment by dgomes** [comment]
You can also add "mediaroom"

**Reply by mib1185** [reply]
Sorry, but I did not get what you mean 🤔

**Reply by dgomes** [reply]
https://www.home-assistant.io/integrations/mediaroom/

It's a media player integration for a set-up-box that also uses `media_next_track` and `media_previous_track` to change channels

**Reply by mib1185** [reply]
ah, I see 👍

**Comment by nick2525** [comment]
I can also add, that my IPTV player APP has next_track button, forward, and next channel. Previous/next track is for time shift. Now, it is very popular on premium IPTV. Some IPTV players can record on HDD disk, it will be also useful to have next_track and next_channel buttons.

**Comment by Drafteed** [comment]
Also suggest to implement `channel_list`, like `source_list`. 

In the current implementation, most integrations store channel lists in the source list, interspersed with a list of applications and inputs. This makes it difficult to fetch the channel list in cases such as [#49676](https://github.com/home-assistant/core/pull/49676). @balloob in this PR says that we should explore that in an architecture issue. If the `media_player` can store the channel list separately, it will give a cleaner way to switch them through the `play_media` service.

**Comment by nick2525** [comment]
Is there any news here?

**Reply by mib1185** [reply]
I'm still waiting for feedback and approval from the maintainers

**Comment by mib1185** [comment]
Hi @bramkragten,
hi @spacegaier
this discussion is a follow up of home-assistant/frontend/pull/8473, those may I ask your opinion on this?

**Reply by mib1185** [reply]
@balloob , @frenck , @MartinHjelmare and @ludeeus your opinion is also likely wanted for this discussion (_and sorry to bother you in this way, but this discussion is already open for more than half a year without response from the core maintainers_) 🙂

**Reply by Drafteed** [reply]
@mib1185 can u also implement channel_list, like source_list (but must contains two parameters: channel number and channel name) and `select_channel` service?

**Reply by Drafteed** [reply]
@mib1185 lg_netcast integration can add to "Suggest integration changes" list too.

**Reply by mib1185** [reply]
> @mib1185 can u also implement channel_list, like source_list (but must contains two parameters: channel number and channel name) and `select_channel` service?

Surely I would support adding `select_channel` service, but honestly not with planed/discussed changes here, since it will not be trivial to implement and those would overwhelm the planed changes.

**Comment by mib1185** [comment]
- home-assistant/frontend/pull/10549
- home-assistant/core/pull/59204

integration specific PRs will follow as soon both PRs above are merged

**Comment by emontnemery** [comment]
While the idea seems perfectly reasonable, it's not clear to me what we gain by adding new services `media_next_channel` / `media_previous_channel`.

What would be the drawback if we instead renamed `media_next_track` and `media_previouos_track` to `media_next` and `media_prev`?
For the frontend button look and feel, is it not enough to check if the `media_content_type` is `MEDIA_TYPE_CHANNEL`?

It's mentioned that most integrations store channel lists in the source list, so a natural follow-up to this improvement would be to add channel list functionality.
Some integrations seem expose the channel  list via the media browsing functionality, for example here: https://github.com/home-assistant/core/pull/46422, does that have drawbacks?

To be clear, I'm looking for a motivation why the existing functionality is not enough and new functionality is needed.

Google does expose the two as separate categories though:
https://developers.google.com/assistant/smarthome/traits/channel 
https://developers.google.com/assistant/smarthome/traits/transportcontrol

**Reply by Drafteed** [reply]
In addition to this, I will say that now it is impossible to determine whether the integration supports channel switching via `play_media`. 
Moreover, integrations that store a list of channels in a source list cause problems such as: 
https://github.com/home-assistant/core/issues/59329
https://github.com/home-assistant/core/pull/59743

> Some integrations seem expose the channel list via the media browsing functionality, for example here: [home-assistant/core#46422](https://github.com/home-assistant/core/pull/46422), does that have drawbacks?

In this case, it is also impossible to determine whether the integration supports channel switching.

**Reply by mib1185** [reply]
> While the idea seems perfectly reasonable, it's not clear to me what we gain by adding new services `media_next_channel` / `media_previous_channel`.

The very first intention was just to bring back the possibility of channel switching, after it was gone with home-assistant/frontend#7389 (already more than a year ago 🙁)
There was also a discussion about this in https://github.com/home-assistant/frontend/pull/8473#issuecomment-826056521 but unfortunately without any consensus 😕 

> What would be the drawback if we instead renamed `media_next_track` and `media_previouos_track` to `media_next` and `media_prev`? For the frontend button look and feel, is it not enough to check if the `media_content_type` is `MEDIA_TYPE_CHANNEL`?

IMO there is a difference between playing/switching a on-demand media and watching/switching a linear TV channel.
Nevertheless, if the `media_content_type` should be used, than it should be clearly defined in the [developer docs](https://developers.home-assistant.io/docs/core/entity/media-player/#mediatype), which media type is used for which purpose.
Btw. what is the intention of the different types `MEDIA_TYPE_TVSHOW` `MEDIA_TYPE_MOVIE` `MEDIA_TYPE_VIDEO` `MEDIA_TYPE_EPISODE` ... a TV-Show or Movie or Episode could be video on demand, but also on linear TV program 🤔 


> It's mentioned that most integrations store channel lists in the source list, so a natural follow-up to this improvement would be to add channel list functionality. Some integrations seem expose the channel list via the media browsing functionality, for example here: [home-assistant/core#46422](https://github.com/home-assistant/core/pull/46422), does that have drawbacks?

As far as i see, this should not be impacted by the planned changes here.
There was also a side-request here to implement some kind of channel list, but this would overwhelm the planned changes here (https://github.com/home-assistant/architecture/discussions/515#discussioncomment-1358916)

> Google does expose the two as separate categories though: https://developers.google.com/assistant/smarthome/traits/channel https://developers.google.com/assistant/smarthome/traits/transportcontrol

👍

**Reply by balloob** [reply]
I think that Erik is onto something. If you're watching a channel, next/prev track will change the channel. 

Watching linear TV should have content type CHANNEL. It should not reflect what the channel is showing.

I noticed in the core PR that there 2 support flags, channel up and down. It should never be 2 flags, you can move the channel or not. 

Also, for channel list discussion -> the idea was to move all those things to the media browser as it can include thumbnails too.

**Reply by mib1185** [reply]
> I think that Erik is onto something. If you're watching a channel, next/prev track will change the channel.

yepp, this is the current implementation in backend, but was refused/removed in frontend with home-assistant/frontend#7389 (_already more than a year ago 🙁_) and with [home-assistant/frontend#8473](https://github.com/home-assistant/frontend/pull/8473#issuecomment-826056521) it was argued that next/prev buttons should only be showed, when player is playing something... but a TV showing a channel is not playing, it is simple on (_most/all TV integrations implement this_)

> Watching linear TV should have content type CHANNEL. It should not reflect what the channel is showing.

i agree with that.

> I noticed in the core PR that there 2 support flags, channel up and down. It should never be 2 flags, you can move the channel or not.

makes absolutely sense!
Maybe we should also consider this for `SUPPORT_NEXT_TRACK` and `SUPPORT_PREVIOUS_TRACK` 🤔 
But `gstreamer` and `pandora` do only implement `SUPPORT_NEXT_TRACK` - all 57 others are implementing both.
In gstreamer it is further curios, since the underlying lib forwards the next command to the stop command 🤨 
The underlying client in pandora seems to really only support next media.
I assume these two integrations may be edge cases 🤔 

---

IMO i would see these two options now:

#### 1) no separate support flags, just use ´MEDIA_TYPE_CHANNEL`
- make sure that TV integrations implement the `MEDIA_TYPE_CHANNEL` properly
- rely on the `SUPPORT_NEXT_TRACK` and `SUPPORT_PREVIOUS_TRACK` flags
- therefore also rely on the `media_next_track()` and `media_previous_track` entity methods and corresponding entity services
- integrations needs to distinguish inside the entity methods if a track or channel should be changed/switched
- show next and prev. track buttons, when "media type is channel" or "player state is playing or paused"
- maybe change icons of these buttons based on situation
  - when media type is channel: use channel switching icons
  - when player state is playing or paused: use media track icons

#### 2) separate support flags, use ´MEDIA_TYPE_CHANNEL`
- make sure that TV integrations implement the `MEDIA_TYPE_CHANNEL` properly
- implement new `SUPPORT_CHANNEL_SWITCH` flag
- implement new `media_next_channel()` and `media_previous_channel` entity methods and corresponding entity services
- (_IMO_) cleaner implementation of next/prev. media and channel switching in integrations
- show channel switching buttons, when media type is channel 
- show next/prev. media track buttons, when player state is playing or paused

**Comment by thecode** [comment]
I would like to try to summarize from the comments above, currently the frontend displays previous/next buttons also when `MEDIA_TYPE_CHANNEL` is active, some integrations use this for overriding the previous/next services behavior with channel down/up. This has two cons:
- Frontend button does not match the operation performed by the backend (a next button is used to send channel up) and impact the user experience.
- Service does not have a fixed behavior. IMO services should have a fixed behavior described by the service name.

From the comments above what we can do which will require minimal changes and will not be a breaking change:
Add a new `supported_feautres` flag `SUPPORT_CHANNEL_SWITCH`, if this flag is set the frontend will display up/down arrow buttons instead of previous/next buttons when media type is `MEDIA_TYPE_CHANNEL` and will call `media_previous_channel` and  `media_next_channel` accordingly.
This does not break any existing integration and let code owners decide if they want to implement this behavihour. It will also provide fixed clear behaviour for services, e.g. `media_next_track` will do a next track command and `media_next_channel` will do a next channel command.

**Comment by mib1185** [comment]
Hi all,

are there any further concerns or suggestions or other comments to https://github.com/home-assistant/architecture/discussions/515#discussioncomment-2065324 and/or https://github.com/home-assistant/architecture/discussions/515#discussioncomment-1725093 ?

**Comment by nick2525** [comment]
any news?
